### ISSUE TITLE
Code: 40. DB::Exception: Checksums of parts don't match: hash of uncompressed files doesn't match. (CHECKSUM_DOESNT_MATCH)

### ISSUE THREAD

**Issue by gauravphoenix** [ClickHouse/ClickHouse#34595]
Pretty typical setup:

- version 22.1.3.
- ingestion through one Kafka table
- just one replicated merge table (2 shards, each shard has three replicas). There is only one MV that moves data from Kafka to the replicated merge table

All nodes are giving these errors and error is shown every one minute.
The only fancy feature enabled is `ratio_of_defaults_for_sparse_serialization = 0.9` in addition to projections

Complete error messge-
```
2022.02.15 00:01:11.349042 [ 313 ] {default.data::ea01da1bfff9c399d15c28ff390ae152_6566_6571_1}
<Error> MergeFromLogEntryTask: Code: 40. DB::Exception: Checksums of parts don't match: hash of uncompressed files doesn't match. (CHECKSUM_DOESNT_MATCH) (version 22.1.3.7 (official build)). 
Data after merge is not byte-identical to data on another replicas. There could be several reasons: 
1. Using newer version of compression library after server update. 2. Using another compression method. 3. Non-deterministic compression algorithm (highly unlikely). 4. Non-deterministic merge algorithm due to logical error in code. 5. Data corruption in memory due to bug in code. 6. Data corruption in memory due to hardware issue. 7. Manual modification of source data after server startup. 8. Manual modification of checksums stored in ZooKeeper. 9. Part format related settings like 'enable_mixed_granularity_parts' are different on different replicas. We will download merged part from replica to force byte-identical result.

```
there are two-three projections that are doing simple aggregates. 

screenshot of errors:

<img width="1357" alt="image" src="https://user-images.githubusercontent.com/627662/153987676-dfa024d4-64bb-4a39-bffc-7f83ec5dd70f.png">

Kafka table settings- 
```
ENGINE = Kafka
SETTINGS kafka_broker_list = '<SERVER_NAME:9092', kafka_topic_list = 'TOPIC_NAME', kafka_group_name = 'ch', kafka_format = 'RowBinary', kafka_skip_broken_messages = 0 
```

**Comment by gauravphoenix** [comment]
@CurtizJ - this is likely due to `ratio_of_defaults_for_sparse_serialization`
when I run `check table` command, I see error: 
```
[] 2022.02.16 22:32:31.711207 [ 993 ] {03a1b2fa-0e1c-4cac-8a2f-b958b873f000} <Error> default.data (ReplicatedMergeTreePartCheckThread): DB::CheckResult DB::ReplicatedMergeTreePartCheckThread::checkPart(const DB::String &): Code: 246. DB::Exception: There is no column <COLUMN_NAME> in serialization infos. (CORRUPTED_DATA), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0xa82d07a in /usr/bin/clickhouse
1. DB::Exception::Exception<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&>(int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&) @ 0x11562ce3 in /usr/bin/clickhouse
2. DB::SerializationInfoByName::readJSON(DB::ReadBuffer&) @ 0x1326df1b in /usr/bin/clickhouse
3. DB::checkDataPart(std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::shared_ptr<DB::IDisk> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::NamesAndTypesList const&, DB::MergeTreeDataPartType const&, std::__1::unordered_set<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, bool, std::__1::function<bool ()>) @ 0x14497708 in /usr/bin/clickhouse
4. DB::checkDataPart(std::__1::shared_ptr<DB::IMergeTreeDataPart const>, bool, std::__1::function<bool ()>) @ 0x1449b7e5 in /usr/bin/clickhouse
5. DB::ReplicatedMergeTreePartCheckThread::checkPart(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0x14445e79 in /usr/bin/clickhouse
6. DB::StorageReplicatedMergeTree::checkData(std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context const>) @ 0x1402fae0 in /usr/bin/clickhouse
7. DB::InterpreterCheckQuery::execute() @ 0x13a13c52 in /usr/bin/clickhouse
8. ? @ 0x13d1434c in /usr/bin/clickhouse

```
where `<COLUMN_NAME>` is a simple `UInt8` in the table. In fact, in the same table, there are other data types like `LowCardinality(String)`, `UInt32 CODEC(DoubleDelta, Default)` etc, but there is only one with `UInt8`
so it is likely something special with `UInt8`

**Comment by gauravphoenix** [comment]
@CurtizJ - so I removed the `ratio_of_defaults_for_sparse_serialization` option from the table and yet I kept getting corruption messages when running `check table` command. After painful debugging, I have been able to track the issue and it seems that it is related to projections. As soon as I remove the projection, I don't see the corruption message. Created a new issue - https://github.com/ClickHouse/ClickHouse/issues/34744
